wp/plugin/lib/cache.ml

```diff
@@ -153,5 +156,20 @@ module Arch = struct
           KB.Conflict.pp c in
       failwith msg
 
+  (* Adds the program's architecture to the KB. *)
+  let save (file : string) (arch : Arch.t) : unit =
```

Comment:
This will break once #1225 is merged. Also, I am not sure I understand why is this needed, can you clarify?

Reply:
We currently have a script that runs wp on every function in a binary with a timeout. This created a large overhead trying to load the binary each time, so we tried saving the program_t and Arch.t into the knowledge base cache. Saving the entire KB to the cache still used up a lot of RAM and time, so we experimented with only saving the program_t and Arch.t while pruning away the other information.

Alternatively, we can try moving the function loop into wp itself, but it seems that we are struggling with getting timeouts working in Ocaml.

Do you have any suggestions?

Reply:
Something is wrong, 
```
        Command being timed: "bap /bin/ls"
        User time (seconds): 2.34
        System time (seconds): 0.41
```
and
```
$ bap cache --info
Capacity:      4096 MB
Current size:    22 MB
```
So storing the full knowledge base of `/bin/ls` takes 22Mb of disk space and the loading time is 2 seconds (less if no objdump or radare2 involved). Can you confirm that you have roughly the same readings, i.e., do

````
bap cache --clean
bap /bin/ls  # warm up the cache
/usr/bin/time -v bap /bin/ls
bap cache --info
```

Reply:
Yes. I agree with these numbers
```
real	0m0.256s
user	0m0.060s
sys	0m0.173s
```

Reply:
In our experience, our tar binaries take 10mins to load the first time. 50s to load the populated complete knowledge base and 5s to load this pruned knowledge base or a saved pruned Program.t

Reply:
Relevant to this discussion:

- https://github.com/draperlaboratory/cbat_tools/issues/245 - An issue discussing how to parrallelize within the WP codebase.
- https://github.com/draperlaboratory/cbat_tools/issues/249 Our current method for running batches
- https://gitter.im/BinaryAnalysisPlatform/bap?at=5f0626978342f4627401cebb A discussion a bit back discussing this using the old cbat save-project plugin

Reply:
```
philip@philip-desktop:$ time bap ../binaries/tar_amd64_linux_gcc_o0_dynamic

real	4m45.780s
user	4m43.940s
sys	0m1.678s
philip@philip-desktop:$ time bap ../binaries/tar_amd64_linux_gcc_o0_dynamic

real	0m8.525s
user	0m7.922s
sys	0m0.528s
```
These numbers are multiplied by at least 2 since we load 2 binaries.

Reply:
What about picking the following strategy. First, disable cache on all stages, i.e., run bap with `--no-cache` option. Second, store only the pruned `program` along with any information that you need (e.g., arch, etc) in a file (not in cache). This will work if you don't perform any computations that touch the knowledge base after you load the program. For example, you shall not create new terms (this will break the tid consistency). If you need this, then I suggest storing the knowledge base and loading it. The numbers do not look that much, I mean 8 seconds to load the project is not that tragic. Let's continue in #245, as I think it is more about parallelizing, rather than about this PR.